modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,348 @@
+use sha2::{Digest, Sha256};
+use subtle_encoding::hex;
+
+use super::error::Error as Ics20Error;
+use crate::applications::transfer::acknowledgement::Acknowledgement;
+use crate::applications::transfer::events::{AckEvent, AckStatusEvent, RecvEvent, TimeoutEvent};
+use crate::applications::transfer::packet::PacketData;
+use crate::applications::transfer::relay::on_ack_packet::process_ack_packet;
+use crate::applications::transfer::relay::on_recv_packet::process_recv_packet;
+use crate::applications::transfer::relay::on_timeout_packet::process_timeout_packet;
+use crate::applications::transfer::{DenomTrace, HashedDenom, IbcCoin, VERSION};
+use crate::core::ics04_channel::channel::{Counterparty, Order};
+use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::Version;
+use crate::core::ics05_port::context::PortReader;
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
+use crate::prelude::*;
+use crate::signer::Signer;
+
+pub trait Ics20Keeper:
+    ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+{
+    type AccountId: ToString;
+
+    /// Sets a new {trace hash -> denom trace} pair to the store.
+    fn set_denom_trace(&mut self, denom_trace: &DenomTrace) -> Result<(), Ics20Error>;
+}
+
+pub trait Ics20Reader:
+    BankReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + AccountReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + ChannelReader
+    + PortReader
+{
+    type AccountId: ToString + TryFrom<Signer>;
```

Comment:
1. Why do we need the `ToString`?
2. Why `TryFrom<Signer>` as opposed to `From<Signer>`? Isn't `Signer` supposed to represent an account as well?
3. `Ics20Keeper::AccountId` only requires `ToString`. Is this intended?

Reply:
1. Nice catch 👌, turns out we don't need `ToString`. (I thought we might need it for logging somewhere but that isn't the case.)
2. An `AccountId` is a valid account type on the host chain while a `Signer` is just a non-empty string with very generic/minimal validation. 
3. Just removed the `ToString` bound all together (a1cdafa27034b0f92e38804239f0a47c9a08426b). The bounds are based on what the lib code uses and since `Ics20Keeper` and its supertraits don't use `TryFrom<Signer>` I didn't add that.